i2c/bmp280_i2c/bmp280_i2c.c

```diff
@@ -0,0 +1,254 @@
+  return converted;
+}
+
+void bmp280_get_calib_params(struct bmp280_calib_param* params) {
+  // raw temp and pressure values need to be calibrated according to
+  // parameters generated during the manufacturing of the sensor
+  // there are 3 temperature params, and 9 pressure params, each with a LSB
+  // and MSB register, so we read from 24 registers
+
+  uint8_t buf[NUM_CALIB_PARAMS] = { 0 };
+  i2c_write_blocking(i2c_default, (addr & BMP280_WRITE_MODE), &REG_DIG_T1_LSB, 1, true);  // true to keep master control of bus
+// read in one go as register addresses auto-increment
+  i2c_read_blocking(i2c_default, (addr & BMP280_READ_MODE), buf, NUM_CALIB_PARAMS, false);  // false, we're done reading
+
+// store these in a struct for later use
+  params->dig_t1 = (uint16_t)(buf[1] << 8) | buf[0];
+  params->dig_t2 = (int16_t)(buf[3] << 8) | buf[2];
+  params->dig_t3 = (int16_t)(buf[5] << 8) | buf[4];
+
+  params->dig_p1 = (uint16_t)(buf[7] << 8) | buf[6];
+  params->dig_p2 = (int16_t)(buf[9] << 8) | buf[8];
+  params->dig_p3 = (int16_t)(buf[11] << 8) | buf[10];
+  params->dig_p4 = (int16_t)(buf[13] << 8) | buf[12];
+  params->dig_p5 = (int16_t)(buf[15] << 8) | buf[14];
+  params->dig_p6 = (int16_t)(buf[17] << 8) | buf[16];
+  params->dig_p7 = (int16_t)(buf[19] << 8) | buf[18];
+  params->dig_p8 = (int16_t)(buf[21] << 8) | buf[20];
+  params->dig_p9 = (int16_t)(buf[23] << 8) | buf[22];
+}
+
+#endif
+
+int main() {
+  stdio_init_all();
+
+  // useful information for picotool
+  bi_decl(bi_2pins_with_func(PICO_DEFAULT_I2C_SDA_PIN, PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C));
+  bi_decl(bi_program_description("BMP280 I2C example for the Raspberry Pi Pico"));
+
+#if !defined(i2c_default) || !defined(PICO_DEFAULT_I2C_SDA_PIN) || !defined(PICO_DEFAULT_I2C_SCL_PIN)
+  #warning i2c / bmp280_i2c example requires a board with I2C pins
+    puts("Default I2C pins were not defined");
+#else
+  printf("Hello, BMP280! Reading temperaure and pressure values from sensor...\n");
+
+  // I2C is "open drain", pull ups to keep signal high when no data is being
+  // sent
+  i2c_init(i2c_default, 100 * 1000);
+  gpio_set_function(PICO_DEFAULT_I2C_SDA_PIN, GPIO_FUNC_I2C);
+  gpio_set_function(PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C);
+  gpio_pull_up(PICO_DEFAULT_I2C_SDA_PIN);
+  gpio_pull_up(PICO_DEFAULT_I2C_SCL_PIN);
+
+  // configure BMP280
+  bmp280_init();
+
+  // retrieve fixed compensation params
+  struct bmp280_calib_param params;
+  bmp280_get_calib_params(&params);
+
+  int32_t raw_temperature;
+  int32_t raw_pressure;
+
+  while (1) {
+    bmp280_read_raw(&raw_temperature, &raw_pressure);
+    float temperature = bmp280_convert_temp(raw_temperature, &params);
+    float pressure = bmp280_convert_pressure(raw_pressure, raw_temperature, &params);
+    printf("Pressure = %.3f kPa\n", pressure / 1000.0);
+    printf("Temp. = %.2f C\n", temperature / 100.0);
+    // poll every 750ms as data refreshes every 500ms
+    sleep_ms(750);
```

Comment:
I don't have a BMP280 myself, and I've not read the datasheet, but is there a good reason for having the sleep-interval in the program different to the data-refresh-interval of the sensor? :shrug:

Reply:
> I don't have a BMP280 myself, and I've not read the datasheet, but is there a good reason for having the sleep-interval in the program different to the data-refresh-interval of the sensor? 🤷

My assumption was so that it didn't hit the refresh boundary, although it might be better not to thave the update time a divisor of the refresh time at that point.

Reply:
Yeah, that's why I asked, since 2 * 0.75 == 3 * 0.5 :wink: 
I guess if you wanted to keep them deliberately non-overlapping you could do something like:
```
start sensor measuring every 500ms
sleep for 250ms
start loop
    read sensor values
    sleep for 500ms
end loop
```
?

Reply:
I just thought that if the sensor updates every 500ms and we poll every 500ms there might be a mismatch in timing. Polling halfway through the update cycle means we always get new data...but I'm not sure

Reply:
But,

500+500+500 = 1,500 ms
750+750 = 1,500 ms

so you still get a sensor mismatch.

Reply:
Ah, that makes..a lot of sense 😅 Changed now.